Add git CET to support all git providers (#2104)
nuclio/nuclio: docs/reference/function-configuration/code-entry-types.md

Set the [`spec.build.codeEntryType`](/docs/reference/function-configuration.md#spec.build.codeEntryType) function-configuration field to one of the following code-entry types to download the function code from the respective external source:

> - `git` &mdash; download the code from a Git repository. See [Git code-entry type (`git`)](#code-entry-type-git).

Reviewer: ```suggestion
- `git` &mdash; clone the code from a Git repository. See [Git code-entry type (`git`)](#code-entry-type-git).
```

> - `github` &mdash; download the code from a GitHub repository. See [GitHub code-entry type (`github`)](#code-entry-type-github). [Deprecated]

Reviewer: deprecate on a following PR, leave this PR purpose to introduce the new CET

> - The downloaded code files are saved and can be used by the function handler.

> **Dashboard Note:** To configure an external function-code source from the dashboard, select the relevant code-entry type &mdash; `Archive`, `Git`, `GitHub`, or `S3` &mdash; from the **Code entry type** list.

The downloaded function code can optionally contain a **function.yaml** file with function configuration for enriching the original configuration (in the configuration file that sets the code-entry type) according to the following merge strategy:

- Field values that are set only in the downloaded configuration are added to the original configuration.
- List and map field values &mdash; such as `meta.labels` and `spec.env` &mdash;are merged by adding any values that are set only in the downloaded configuration to the values that are set in the original configuration.
- In case of a conflict &mdash; i.e., if the original and downloaded configurations set different values for the same element &mdash; the original configuration takes precedence and the value in the downloaded configuration is ignored.

<a id="code-entry-type-git"></a>
### Git code-entry type (`git`)

> Set the [`spec.build.codeEntryType`](/docs/reference/function-configuration.md#spec.build.codeEntryType) function-configuration field to `git` (dashboard: **Code entry type** = `Git`) to download the function code from a Git repository. The following configuration fields provide additional information for performing the download:

Reviewer: ```suggestion
Set the [`spec.build.codeEntryType`](/docs/reference/function-configuration.md#spec.build.codeEntryType) function-configuration field to `git` (dashboard: **Code entry type** = `Git`) to clone the function code from a Git repository. The following configuration fields provide additional information for performing the cloning:
```

- `spec.build` &mdash;
  - `path` (dashboard: **URL**) (Required) &mdash; the URL of the Git repository that contains the function code.
  - `codeEntryAttributes` &mdash;
      - `branch` (dashboard: **Branch**) (Required) &mdash; the Git repository branch from which to download the function code.

>       - `gitCredentials` (Optional) &mdash; Git credentials for the download, when required.

Reviewer: ```suggestion
      - `gitCredentials` (Optional) &mdash; Git credentials to use for authentication.
```

>         - `accessToken` (dashboard: **Token**) (Optional) Git access key

Reviewer: what is `accessToken` and how it is being used?

Imo, all repos support PAT (personal access token). each repo uses it differently.
e.g.:
- `AzureDevops` on `password` field.
- `Github` on `username` field

I'd suggest to leave `username/password` for now. it is up to the user to understand what his git server expects to see.

        - `username` (dashboard: **Username**) (Optional) Git username
        - `password` (dashboard: **Password**) (Optional) Git password
      - `workDir` (dashboard: **Work directory**) (Optional) &mdash; the relative path to the function-code directory within the configured repository branch.
      The default work directory is the root directory of the Git repository (`"/"`).

<a id="code-entry-type-git-example"></a>
#### Example

```yaml
spec:
  description: my Go function
  handler: main:Handler
  runtime: golang
  build:
    codeEntryType: "git"
    path: "https://git.com/my-organization/my-repository.git"
    codeEntryAttributes:
      branch: "my-branch"
      gitCredentials:

>         accessKey: "myaccesskey"

Reviewer: imo, provide examples for some of the common repositories
`AzureDevops`
`Github`
`BitBucket`

it can be very helpful. feel free to add a specific CET reference and nest it under `codeEntryType` dir.
there we can add more custom examples for ease and explain its usage to the user.
this feature it going to be used by a lot of people, so more examples == less issues :)

        username: "myusername"
      workDir: "/go/myfunc"
```

<a id="code-entry-type-github"></a>

> ### GitHub code-entry type (`github`) [Deprecated]

Reviewer: deprecate on another PR
